README
#==============================================================
#
#  Name: README
#
#  Rem:  This is a driver for the Acromag IP231 16-channel 
#        Industry Pack (IP) analog output card. 
#
#  Auth: 29-Aug-2007, Sheng Peng         (PENGS)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        dd-mmm-yyyy, First Lastname     (USERNAME):
#          comment
#
#==============================================================
#
EPICS Verions:
--------------
This driver has currently been built and tested at SLAC with EPICS 3.14.8.2.
This driver has implemented EPICS OSI function calls and is OSI compliant.

Operating System:
-------------------
This driver has only been tested  under RTEMS. 

How to use this module in your IOC
------------------
1.	Add the following lines in your iocTop/configure/RELEASE:
IP231=$(EPICS_MODULES)/ip231/$(IP231_MODULE_VERSION)

2.	Copy the following files to your iocTop/<iocName>/<iocName>App/Db directory:
	ip231App/Db/IP231.substitutions

3.	Add the following lines in your iocTop/<iocName>/<iocName>App/Db/Makefile:
DB += IP231.db

4.	Add the following lines in your iocTop/<iocName>/<iocName>App/src/Makefile:
<iocName>_DBD += IP231.dbd
<iocName>_LIBS += IP231

5.	Add the following lines in your iocTop/<iocName>/<iocName>App/src/<iocName>Main.cpp:
//  Include these lines to force linker to include xy9660 Ipac carrier
//	Note: There are other ipac carrier modules besides the xy9660.
//	See the ipac module for options.
#include "drvXy9660.h"
extern "C"
int addCarrier( const char * cardParams )
{
    return ipacAddXVME9660( cardParams );
}

6.	Add the following lines in your iocTop/<iocName>/iocBoot/ioc<iocName>/st.cmd:
# Initialize IPAC carrier
# The 2nd parameter is the base I/O address set by the jumpers on your carrier board
# Make sure this address doesn't conflict with other boards.
addCarrier( "0x0000" )

# Initialize IP231 analog output module
# ip231Create( <name>, <card>, <slot>, <dacmode> )
# <card> is 0 for the first ipac carrier
# <slot> is 0..3 for carrier slots A..D
# <dacmode> must be "transparent" or "simultaneous"
# In transparent mode, outputs are updated when written
# In simultaneous mode, multiple output values can be setup,
# and all outputs update when ip231SimulTrigger(<cardNo>) is called
ip231Create( "ao0", 0, 3, "transparent" )

# Load the EPICS records
dbLoadRecords("db/IP231.db","PRE=TST:AOUT:,CARD=ao0")

############################################################

EPICS PV Values:
----------------------
The default names assigned by the db template and substitution files
in this module are shown below.  These can be customized after you copy the
template and substitution files to your ioc.
------------------
PV Name:	Type:		Description:
TST:AOUT:0	ao			Analog Output 0
TST:AOUT:1	ao			Analog Output 1
TST:AOUT:2	ao			Analog Output 2
TST:AOUT:3	ao			Analog Output 3
TST:AOUT:4	ao			Analog Output 4
TST:AOUT:5	ao			Analog Output 5
TST:AOUT:6	ao			Analog Output 6
TST:AOUT:7	ao			Analog Output 7
TST:AOUT:8	ao			Analog Output 8
TST:AOUT:9	ao			Analog Output 9
TST:AOUT:10	ao			Analog Output 10
TST:AOUT:11	ao			Analog Output 11
TST:AOUT:12	ao			Analog Output 12
TST:AOUT:13	ao			Analog Output 13
TST:AOUT:14	ao			Analog Output 14
TST:AOUT:15	ao			Analog Output 15


RELEASE_NOTES
#==============================================================
#
#  Name: RELEASE_NOTES
#
#  Rem: This files describes release notes for
#       the Acromag ip231 driver and device support.      
#
#  Auth: 29-Aug-2007, Sheng Peng         (PENGS)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        21-Feb-2008, K. Luchini         (LUCHINI):
#          release ip231-R1-0-5
#        10-Dec-2007, K. Luchini         (LUCHINI):
#          release ip231-R1-0-4
#        26-Sep-2007, K. Luchini         (LUCHINI):
#          release ip231-R1-0-3
#          release ip231-R1-0-2
#        30-Aug-2007, K. Luchini         (LUCHINI):
#          release ip231-R1-0-1
#
#==============================================================
#
R2.2.3-1.5.3   17-Oct-2022,  V. Esposito (espov):
               On R2.2.3-1.branch
               Point to latest IPAC release, compile for rhel7

R2.3.0-0.2.3   12-Aug-2022,  K. Luchini (luchini):
               On R2.3.0-0.branch
               Add build for linuxRT-x86-64

R2.3.0-0.2.2   21-Apr-2022,  K. Luchini (luchini):
               On R2.3.0-0.branch
               add macro DESC with default
               update ipac module to R2.15-1.0.0

R2.2.3-1.5.0    build for epics 7
                align with ipac R2.14-0.2.0
                use typed RSET and type DRVET
                include errlog.h to avoid undefined symbol for errlogPrintf
                bug fix for EPICS_VERSION check

R2.2.3-1.4.2    Changing to support some of the git workflow
                Remove autosave and archive files

ip231-R1-0-5:   Build with base-R3-14-9-lcls1 (RTEMS 4.7.1)
                Build with ipac-R2-8-lcls4  
                Replace CONFIG_APP with CONFIG_SITE
                Add RELEASE_SITE  

ip231-R1-0-4:   Backout epics version to bas-R3-14-8-2
                Build with ipac-R2-8-lcls3 w/no hotswap 
                of the hytec 8002 ip carrier

ip231-R1-0-3:   Build with base-R3-14-9-lcls1 (RTEMS 4.7.1)
                Build with ipac-R2-8-lcls2

ip231-R1-0-2:   Build with base-R3-14-8-2-lcls2 (RTEMS 4.7.1)
                Build with ipac-R2-8-lcls1

ip231-R1-0-1:   Build with ipac-R1-0-2 instead of ipac-R1-0-1
                Built with base-R3-14-8-2-lcls1 (RTEMS 4.7)

ip231-R1-0-0:   Initial release


ip231App/src/devAoIP231.c
/****************************************************************/
/* $Id: devAoIP231.c,v 1.2 2006/08/10 16:28:28 pengs Exp $      */
/* This file implements AO record device support for IP231 DAC  */
/* Author: Sheng Peng, pengs@slac.stanford.edu, 650-926-3847    */
/****************************************************************/
#include <stdio.h>
#include <string.h>

#include <epicsVersion.h>

#if (EPICS_VERSION>=3 && EPICS_REVISION>=14) || EPICS_VERSION>3
#include <epicsExport.h>
#endif

#include <errlog.h>
#include <devLib.h>
#include <dbAccess.h>
#include <dbScan.h>
#include <callback.h>
#include <cvtTable.h>
#include <link.h>
#include <recSup.h>
#include <recGbl.h>
#include <devSup.h>
#include <drvSup.h>
#include <dbCommon.h>
#include <alarm.h>
#include <cantProceed.h>
#include <aoRecord.h>

#include <ptypes.h>
#include <drvIP231Lib.h>

#define MAX_CA_STRING_SIZE (40)

/* define function flags */
typedef enum {
        IP231_AO_DATA,
} IP231FUNC;

static struct PARAM_MAP
{
        char param[MAX_CA_STRING_SIZE];
        int  funcflag;
} param_map[1] = {
    {"DATA", IP231_AO_DATA}
};
#define N_PARAM_MAP (sizeof(param_map)/sizeof(struct PARAM_MAP))

typedef struct IP231_DEVDATA
{
    IP231_ID	pcard;
    UINT16	chnlnum;
    int		funcflag;
} IP231_DEVDATA;

/* This function will be called by all device support */
/* The memory for IP231_DEVDATA will be malloced inside */
static int IP231_DevData_Init(dbCommon * precord, char * ioString)
{
    int		count;
    int		loop;

    char	cardname[MAX_CA_STRING_SIZE];
    IP231_ID	pcard;
    int		chnlnum;
    char	param[MAX_CA_STRING_SIZE];
    int		funcflag = 0;

    IP231_DEVDATA *   pdevdata;

    /* param check */
    if(precord == NULL || ioString == NULL)
    {
        if(!precord) errlogPrintf("No legal record pointer!\n");
        if(!ioString) errlogPrintf("No INP/OUT field for record %s!\n", precord->name);
        return -1;
    }

    /* analyze INP/OUT string */
    count = sscanf(ioString, "%[^:]:%i:%[^:]", cardname, &chnlnum, param);
    if (count != 3)
    {
        errlogPrintf("Record %s INP/OUT string %s format is illegal!\n", precord->name, ioString);
        return -1;
    }

    pcard = ip231GetByName(cardname);
    if( !pcard )
    {
        errlogPrintf("Record %s IP231 %s is not registered!\n", precord->name, cardname);
        return -1;
    }

    if(chnlnum < 0 || chnlnum >= 16 )
    {/* chnlnum is UINT16, more accurate check again start/end channel will be done in ip231Read/ip231Write */
        errlogPrintf("Record %s channel number %d is out of range for IP231 %s!\n", precord->name, chnlnum, cardname);
        return -1;
    }

    for(loop=0; loop<N_PARAM_MAP; loop++)
    {
        if( 0 == strcmp(param_map[loop].param, param) )
        {
            funcflag = param_map[loop].funcflag;
            break;
        }
    }
    if(loop >= N_PARAM_MAP)
    {
        errlogPrintf("Record %s param %s is illegal!\n", precord->name, param);
        return -1;
    }

    pdevdata = (IP231_DEVDATA *)callocMustSucceed(1, sizeof(IP231_DEVDATA), "Init record for IP231");

    pdevdata->pcard = pcard;
    pdevdata->chnlnum = chnlnum;
    pdevdata->funcflag = funcflag;

    precord->dpvt = (void *)pdevdata;
    return 0;
}


static long init_ao( struct aoRecord * pao)
{
    IP231_DEVDATA * pdevdata;
    signed int temp;

    pao->dpvt = NULL;

    if (pao->out.type!=INST_IO)
    {
        recGblRecordError(S_db_badField, (void *)pao, "devAoIP231 Init_record, Illegal INP");
        pao->pact=TRUE;
        return (S_db_badField);
    }
    pao->eslo = (pao->eguf - pao->egul)/(float)0x10000;
    pao->roff = 0x0;

    if(IP231_DevData_Init((dbCommon *) pao, pao->out.value.instio.string) != 0)
    {
        errlogPrintf("Fail to init devdata for record %s!\n", pao->name);
        recGblRecordError(S_db_badField, (void *) pao, "Init devdata Error");
        pao->pact = TRUE;
        return (S_db_badField);
    }

    pdevdata = (IP231_DEVDATA *)(pao->dpvt);
    if( 0 == ip231Read(pdevdata->pcard, pdevdata->chnlnum, &temp))
    {
        pao->rval = temp;
        pao->udf = FALSE;
        pao->stat = pao->sevr = NO_ALARM;
        return 0; /* convert */
    }
    else
    {
        return -1;
    }
}

static long write_ao(struct aoRecord *pao)
{
    IP231_DEVDATA * pdevdata = (IP231_DEVDATA *)(pao->dpvt);

    int status=-1;
    signed int tmp;

    switch(pdevdata->funcflag)
    {
    case IP231_AO_DATA:
        tmp = pao->rval;
        status = ip231Write(pdevdata->pcard, pdevdata->chnlnum, tmp);
        break;
    }

    if(status)
    {
        recGblSetSevr(pao, WRITE_ALARM, INVALID_ALARM);
        return -1;
    }

    return 0;
}

static long ao_lincvt(struct aoRecord   *pao, int after)
{

        if(!after) return(0);
        /* set linear conversion slope*/
        pao->eslo = (pao->eguf - pao->egul)/(float)0x10000;
        pao->roff = 0x0;
        return(0);
}

struct IP231_DEV_SUP_SET
{
    long            number;
    DEVSUPFUN       report;
    DEVSUPFUN       init;
    DEVSUPFUN       init_record;
    DEVSUPFUN       get_ioint_info;
    DEVSUPFUN       write_ao;
    DEVSUPFUN       special_linconv;
} devAoIP231 = {6, NULL, NULL, init_ao, NULL, write_ao, ao_lincvt};

#if (EPICS_VERSION>=3 && EPICS_REVISION>=14) || EPICS_VERSION>3
epicsExportAddress(dset, devAoIP231);
#endif



ip231App/src/drvIP231Lib.h
#ifndef _INCLUDE_DRV_IP231_LIB_H_
#define _INCLUDE_DRV_IP231_LIB_H_

/****************************************************************/
/* $Id: drvIP231Lib.h,v 1.2 2006/08/10 16:27:35 pengs Exp $     */
/* This file defines the user interface of IP231 support        */
/* Author: Sheng Peng, pengs@slac.stanford.edu, 650-926-3847    */
/****************************************************************/

#ifdef __cplusplus
extern "C" {
#endif  /* __cplusplus */

#include "ptypes.h"

typedef struct IP231_CARD * IP231_ID;

int ip231Create (char *cardname, UINT16 carrier, UINT16 slot, char *dacmode);


IP231_ID ip231GetByName(char * cardname);
IP231_ID ip231GetByLocation(UINT16 carrier, UINT16 slot);

int ip231Write(IP231_ID pcard, UINT16 channel, signed int value);
int ip231Read(IP231_ID pcard, UINT16 channel, signed int * pvalue);

void ip231SimulTrigger(IP231_ID pcard);
void ip231SimulTriggerByName(char * cardname);

#ifdef __cplusplus
}
#endif  /* __cplusplus */


#endif


ip231App/src/drvIP231Private.h
#ifndef _INCLUDE_DRV_IP231_PRIVATE_H_
#define _INCLUDE_DRV_IP231_PRIVATE_H_

#ifdef __cplusplus
extern "C" {
#endif  /* __cplusplus */

/******************************************************************/
/* $Id: drvIP231Private.h,v 1.3 2006/09/06 23:43:37 pengs Exp $   */
/* This file defines the internal hw/sw struct of IP231 DAC module*/
/* Author: Sheng Peng, pengs@slac.stanford.edu, 650-926-3847      */
/******************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include <epicsVersion.h>
#if (EPICS_VERSION>=3 && EPICS_REVISION>=14) || EPICS_VERSION>3

#include <epicsMutex.h>
#include <epicsThread.h>
#include <epicsString.h>
#include <epicsInterrupt.h>
#include <cantProceed.h>
#include <epicsExport.h>
#include <drvSup.h>
#include <dbScan.h>
#include <ellLib.h>

#include "drvIpac.h"

#include "ptypes.h"

#else
#error "You need EPICS 3.14 or above because we need OSI support!"
#endif

#define IP231_DRV_VERSION "IP231 driver V1.0"

/* Manufacturers assigned values */
#define IP_MANUFACTURER_ACROMAG         0xa3
#define IP_MODEL_DAC_231_16             0x34
#define IP_MODEL_DAC_231_8              0x33

#define MAX_IP231_16_CHANNELS           16
#define MAX_IP231_8_CHANNELS            8

#define CTRL_REG_RESET                  0x80

#define DAC_MODE_TRANS			0x0
#define DAC_MODE_SIMUL			0x1

#define DEBUG_MSG_SIZE 256

/* Hardware registers for DAC channels */
typedef struct IP231_HW_MAP
{
    volatile UINT16   data[MAX_IP231_16_CHANNELS];	/* DAC channel data. (R/W), the last 8 is only available to 16 channel version */
    volatile UINT16   transMode;	/* transparent mode (W), value doesn't matter */
    volatile UINT16   simulMode;	/* simultaneous mode (W), value doesn't matter */
    volatile UINT16   simulTrig;	/* simultaneous output trigger (W), value doesn't matter */
    volatile UINT16   writeStatus;	/* DAC write status reg. (R) Each bit "0" indicates busy for each channel */
    volatile UINT16   controlReg;	/* Control Register (LSB) (R/W), bit 7 for reset */

    volatile UINT8    eepromStatus;	/* EEPROM status resister (R), bit 7 "0" indicates eeprom busy */
    volatile UINT8    eepromControl;	/* EEPROM status resister (R/W), bit 0,1 to enable write to eeprom */

    volatile UINT16   unUsed[10];	/* unused words */
    volatile UINT8    calData[64];	/* gain and offset, D8 access only,(R)  */
} IP231_HW_MAP;
/* We don't use __attribute__ ((packed)) here because it is already naturally aligned */
/* And packed with -mstrict-align will make access to byte-access, it will hurt */


/* device driver ID structure */

typedef ELLLIST IP231_CARD_LIST;

typedef struct IP231_CARD
{
    ELLNODE                     node;		/* Link List Node */

    epicsMutexId                lock;

    char                        * cardname;	/* Card identification */
    UINT16                      carrier;	/* Industry Pack Carrier Index */
    UINT16                      slot;		/* Slot number on carrier */

    UINT32                      num_chnl;	/* Number of Channels, 16 or 8 depend on model */ 

    UINT32                      dac_mode;	/* transparent mode or simultaneous mode */

    volatile IP231_HW_MAP       *pHardware;	/* controller registers */

    double                      adj_offset[MAX_IP231_16_CHANNELS];
    double                      adj_slope[MAX_IP231_16_CHANNELS];

    char                        debug_msg[DEBUG_MSG_SIZE];
} IP231_CARD;


#ifdef __cplusplus
}
#endif  /* __cplusplus */

#endif


ip231App/src/drvIP231.c
/****************************************************************/
/* $Id: drvIP231.c,v 1.1.1.1 2006/08/10 16:31:19 luchini Exp $        */
/* This file implements driver support for IP231 DAC            */
/* Author: Sheng Peng, pengs@slac.stanford.edu, 650-926-3847    */
/****************************************************************/

#ifndef NO_EPICS
#include "devLib.h"
#include "drvSup.h"
#include "dbScan.h"
#include "epicsInterrupt.h"
#include "epicsExport.h"
#include "iocsh.h"
#endif

#include "errlog.h"
#include "drvIP231Lib.h"
#include "drvIP231Private.h"

int    IP231_DRV_DEBUG = 0;
int    IP231_DRV_DEBUG_CH = -1;

static IP231_CARD_LIST	ip231_card_list;
static int		card_list_inited=0;

/*****************************************************************/
/* Find IP231_CARD which matches the cardname from link list     */
/*****************************************************************/
IP231_ID ip231GetByName(char * cardname)
{
    IP231_ID pcard = NULL;

    if(!card_list_inited)   return NULL;

    for(pcard=(IP231_ID)ellFirst((ELLLIST *)&ip231_card_list); pcard; pcard = (IP231_ID)ellNext((ELLNODE *)pcard))
    {
        if ( 0 == strcmp(cardname, pcard->cardname) )   break;
    }

    return pcard;
}

/*****************************************************************/
/* Find IP231_CARD which matches the carrier/slot from link list */
/*****************************************************************/
IP231_ID ip231GetByLocation(UINT16 carrier, UINT16 slot)
{
    IP231_ID pcard = NULL;

    if(!card_list_inited)   return NULL;

    for(pcard=(IP231_ID)ellFirst((ELLLIST *)&ip231_card_list); pcard; pcard = (IP231_ID)ellNext((ELLNODE *)pcard))
    {
        if ( (carrier == pcard->carrier) && (slot == pcard->slot) )   break;
    }

    return pcard;
}

/***************************************************************************************************************************/
/*  Routine: ip231Create                                                                                                   */
/*                                                                                                                         */
/*  Purpose: Register a new IP-231 DAC module                                                                              */
/*                                                                                                                         */
/*  Description:                                                                                                           */
/*                                                                                                                         */
/*  Checks all parameters, then creates a new card structure, initializes it and adds it to the end of the linked list.    */
/*                                                                                                                         */
/*  SYNOPSIS: int ip231Create(                                                                                             */
/*   		    char *cardname,	  Unique Identifier "ip231-1"                                                      */
/*    		    UINT16 carrier,	  Ipac Driver carrier card number                                                  */
/*   		    UINT16 slot,	  Slot number on IP carrier                                                        */
/*                  char *dacmode)        This must be "transparent" or "simultaneous"                                     */
/*  Example:                                                                                                               */
/*            ip231Create("ip231_1", 0, 0, "transparent")                                                                  */
/***************************************************************************************************************************/

int ip231Create (char *cardname, UINT16 carrier, UINT16 slot, char *dacmode)
{
    int status=0, loop;

    /* calloc will put everything zero */
    IP231_ID pcard = callocMustSucceed(1, sizeof(struct IP231_CARD), "ip231Create");
   
    if(!card_list_inited)
    {/* Initialize the IP231 link list */
        ellInit( (ELLLIST *) &ip231_card_list);
        card_list_inited = 1;
        if(IP231_DRV_DEBUG) printf("The size of IP231_HW_MAP is %zu\n", sizeof(struct IP231_HW_MAP));
    }

    pcard->lock = epicsMutexMustCreate();

    /************************************ Parameters check ************************************/

    /* Check cardname */
    if( (!cardname) || (0 == strlen(cardname)) )
    {
        errlogPrintf ("ip231Create: No cardname specified!\n");
        status = -1;
        goto FAIL;
    }
    if( ip231GetByName(cardname) )
    {
        errlogPrintf ("ip231Create: %s already existed!\n", cardname);
        status = -1;
        goto FAIL;
    }
    pcard->cardname = epicsStrDup(cardname);

    /* Check if we really have the Acromag IP231 installed on carrier/slot */
    if(0 == ipmValidate(carrier, slot, IP_MANUFACTURER_ACROMAG, IP_MODEL_DAC_231_16))
    {
        pcard->num_chnl = MAX_IP231_16_CHANNELS;
    }
    else if(0 == ipmValidate(carrier, slot, IP_MANUFACTURER_ACROMAG, IP_MODEL_DAC_231_8))
    {
        pcard->num_chnl = MAX_IP231_8_CHANNELS;
    }
    else
    {
        errlogPrintf ("ip231Create: Carrier %d Slot %d has no IP231\n", carrier, slot);
        status = -1;
        goto FAIL;
    }
    if( ip231GetByLocation(carrier, slot) )
    {
        errlogPrintf ("ip231Create: IP231 on carrier %d slot %d is already registered!\n", carrier, slot);
        status = -1;
        goto FAIL;
    }
    /* Slot contains a real IP-231 module */
    pcard->carrier = carrier;
    pcard->slot = slot;

    /* Check DAC mode */
    if(0 == strcmp(dacmode, "transparent"))
    {
        pcard->dac_mode = DAC_MODE_TRANS;
    }
    else if(0 == strcmp(dacmode, "simultaneous"))
    {
        pcard->dac_mode = DAC_MODE_SIMUL;
    }
    else
    {
        errlogPrintf ("ip231Create: dacmode %s is illegal for device %s\n", dacmode, cardname);
        status = -1;
        goto FAIL;
    }

    /* Hardware pointer */
    pcard->pHardware   = (IP231_HW_MAP *) ipmBaseAddr(carrier, slot, ipac_addrIO);

    /* Read EEPROM to get calibration data */
    for(loop = 0; loop < pcard->num_chnl; loop++)
    {
        UINT16 offset, gain;
        SINT16 offset_val, gain_val;

        offset = pcard->pHardware->calData[loop*4+0];
        gain = pcard->pHardware->calData[loop*4+2];  

        if(IP231_DRV_DEBUG) printf("Card %s, Channel %d, offset_err 0x%04x, gain_err 0x%04x\n", cardname, loop, offset, gain);

        offset_val = (SINT16)offset;
        gain_val = (SINT16)gain;
        pcard->adj_slope[loop] = 1.0 + (double)gain_val/262144.0;
        pcard->adj_offset[loop] = (double)offset_val/4.0 - (double)gain_val/8.0;
        if(IP231_DRV_DEBUG) printf("Card %s, Channel %d, adj_slope %g, adj_offset %g\n", cardname, loop, pcard->adj_slope[loop], pcard->adj_offset[loop]);

    }

    /* We don't want to reset, but even soft reboot reset bus and reset IP231, so we just reset again */
    pcard->pHardware->controlReg = CTRL_REG_RESET;
    epicsThreadSleep(0.1);
    for(loop = 0; loop < pcard->num_chnl; loop++)
    {
        ip231Write(pcard, loop, 0x8000);
    }
    epicsThreadSleep(0.1);

    if(pcard->dac_mode == DAC_MODE_TRANS)
    {/* Transparent mode */
        pcard->pHardware->transMode = 0xFFFF;
    }
    else
    {/* Simultaneous mode */
        pcard->pHardware->simulMode = 0xFFFF;
    }

    /* We successfully allocate all resource */
    ellAdd( (ELLLIST *)&ip231_card_list, (ELLNODE *)pcard);

    return 0;

FAIL:
    if(pcard->lock) epicsMutexDestroy(pcard->lock);
    if(pcard->cardname) free(pcard->cardname);
    free(pcard);
    return status;
}

/****************************************************************/
/* Write data to paticular channel, do correction first         */
/****************************************************************/
int ip231Write(IP231_ID pcard, UINT16 channel, signed int value)
{
    /* The raw value from AO record is 0 ~ 65535 */
    /* But after aslo/aoff, it might be a little bit wider range */

    signed int tmp;
    UINT16 dac_val;
    UINT32 mask;

    if(!pcard)
    {
        errlogPrintf("ip231Write called with NULL pointer!\n");
        return -1;
    }

    if(channel >= pcard->num_chnl)
    {
        errlogPrintf("Bad channel number %d in ip231Write card %s\n", channel, pcard->cardname);
        return -1;
    }

    if ( IP231_DRV_DEBUG && (IP231_DRV_DEBUG_CH < 0 || IP231_DRV_DEBUG_CH == channel) )
		printf("Write %d to card %s channel %d\n", value, pcard->cardname, channel);

    tmp = value * pcard->adj_slope[channel] + pcard->adj_offset[channel];
    if(tmp > 65535)
        dac_val = 65535;
    else if(tmp < 0)
        dac_val = 0;
    else
        dac_val = tmp;

    epicsMutexLock(pcard->lock);

    mask = (0x1 << channel);

    while( !(mask & (pcard->pHardware->writeStatus)) );

    if ( (IP231_DRV_DEBUG > 1) && (IP231_DRV_DEBUG_CH < 0 || IP231_DRV_DEBUG_CH == channel) )
		printf("Write corrected %d (%#04x) to card %s channel %d@%p\n", dac_val, dac_val, pcard->cardname, channel, &(pcard->pHardware->data[channel]));

    pcard->pHardware->data[channel] = dac_val;

    epicsMutexUnlock(pcard->lock);

    return 0;
}

/****************************************************************/
/* Read data from paticular channel, do reverse-correction first*/
/****************************************************************/
int ip231Read(IP231_ID pcard, UINT16 channel, signed int * pvalue)
{
    /* The raw value from AO record is 0 ~ 65535 */
    /* But after aslo/aoff, it might be a little bit wider range */

    UINT16 dac_val;
    UINT32 mask;

    if(!pcard)
    {
        errlogPrintf("ip231Read called with NULL pointer!\n");
        return -1;
    }

    if(channel >= pcard->num_chnl)
    {
        errlogPrintf("Bad channel number %d in ip231Read card %s\n", channel, pcard->cardname);
        return -1;
    }

    epicsMutexLock(pcard->lock);

    mask = (0x1 << channel);

    while( !(mask & (pcard->pHardware->writeStatus)) );
    dac_val = pcard->pHardware->data[channel];

    epicsMutexUnlock(pcard->lock);

    *pvalue = (dac_val - pcard->adj_offset[channel])/pcard->adj_slope[channel];

    return 0;
}

/***********************************************************************/
/* Simultaneous trigger                                                */
/***********************************************************************/
void ip231SimulTrigger(IP231_ID pcard)
{
    if(pcard)
    {
        epicsMutexLock(pcard->lock);
        if(pcard->dac_mode == DAC_MODE_SIMUL)
            pcard->pHardware->simulTrig = 0xFFFF;
        epicsMutexUnlock(pcard->lock);
    }
}

void ip231SimulTriggerByName(char * cardname)
{
    IP231_ID pcard = ip231GetByName(cardname);
    ip231SimulTrigger(pcard);
}

/**************************************************************************************************/
/* Here we supply the driver report function for epics                                            */
/**************************************************************************************************/
static  long    IP231_EPICS_Report(int level);

#ifndef USE_TYPED_DRVET
const struct drvet drvIP231 = {2,                              /*2 Table Entries */
                              (DRVSUPFUN) IP231_EPICS_Report,  /* Driver Report Routine */
                              NULL}; /* Driver Initialization Routine */
#else
const drvet drvIP231 = {2, IP231_EPICS_Report, NULL};
#endif

epicsExportAddress(drvet,drvIP231);

/* implementation */
static long IP231_EPICS_Report(int level)
{
    IP231_ID pcard;

    printf("\n"IP231_DRV_VERSION"\n\n");

    if(!card_list_inited)
    {
        printf("IP231 card link list is not inited yet!\n\n");
        return 0;
    }

    if(level > 0)   /* we only get into link list for detail when user wants */
    {
        for(pcard=(IP231_ID)ellFirst((ELLLIST *)&ip231_card_list); pcard; pcard = (IP231_ID)ellNext((ELLNODE *)pcard))
        {
            printf("\tIP231 card %s is installed on carrier %d slot %d\n", pcard->cardname, pcard->carrier, pcard->slot);

            if(level > 1)
            {
                printf("\tIn total %d channels, DAC mode is %s\n", pcard->num_chnl, (pcard->dac_mode)==DAC_MODE_SIMUL?"Simultaneous":"Transparent");
                printf("\tIO space is at %p\nn", pcard->pHardware);
            }
        }
    }

    return 0;
}
/*******************************************************************************
* EPICS iocsh Command registry
*/

#ifndef NO_EPICS

#if     0
/* ip231Report(int interest) */
static const iocshArg ip231ReportArg0 = {"interest", iocshArgInt};
static const iocshArg * const ip231ReportArgs[1] = {&ip231ReportArg0};
static const iocshFuncDef ip231ReportFuncDef =
    {"ip231Report",1,ip231ReportArgs};
static void ip231ReportCallFunc(const iocshArgBuf *args)
{
    ip231Report(args[0].ival);
}
#endif


/* ip231Create( char *pName, unsigned short card, unsigned short slot, char *modeNamer); */

static const iocshArg ip231CreateArg0 = {"pName",iocshArgPersistentString};
static const iocshArg ip231CreateArg1 = {"card", iocshArgInt};
static const iocshArg ip231CreateArg2 = {"slot", iocshArgInt};
static const iocshArg ip231CreateArg3 = {"modeName",iocshArgString};

static const iocshArg * const ip231CreateArgs[9] = {
    &ip231CreateArg0, &ip231CreateArg1, &ip231CreateArg2, &ip231CreateArg3};

static const iocshFuncDef ip231CreateFuncDef =
    {"ip231Create",4,ip231CreateArgs};

static void ip231CreateCallFunc(const iocshArgBuf *arg)
{
    ip231Create(arg[0].sval, arg[1].ival, arg[2].ival, arg[3].sval);
}

LOCAL void drvIP231Registrar(void) {
/*    iocshRegister(&ip231ReportFuncDef,ip231ReportCallFunc); */
    iocshRegister(&ip231CreateFuncDef,ip231CreateCallFunc);
}
epicsExportRegistrar(drvIP231Registrar);
epicsExportAddress( int, IP231_DRV_DEBUG );
epicsExportAddress( int, IP231_DRV_DEBUG_CH );

#endif /* NO_EPICS */

